grease/src/Grease/Refine.hs

```diff
@@ -568,6 +555,81 @@ execAndRefine bak solver _fm la memVar anns heuristics argNames argShapes initSt
       ProveCantRefine (Timeout{}) -> "symex timeout"
       ProveCantRefine (Unsupported{}) -> "unsupported feature"
 
+-- TODO: Fold into `refinementLoop`?
+refineOnce ::
+  ( Mem.HasPtrWidth wptr
+  , C.IsSyntaxExtension ext
+  , OnlineSolverAndBackend solver sym bak t st (W4.Flags fm)
+  , ToConc.HasToConcretize p
+  , ?memOpts :: Mem.MemOptions
+  , ExtShape ext ~ PtrShape ext wptr
+  , Cursor.CursorExt ext ~ PtrCursor.Dereference ext wptr
+  ) =>
+  GreaseLogAction ->
+  Opts.SimOpts ->
+  C.HandleAllocator ->
+  bak ->
+  W4.FloatModeRepr fm ->
+  DataLayout ->
+  Ctx.Assignment Setup.ValueName argTys ->
+  Ctx.Assignment (Const String) argTys ->
+  Ctx.Assignment C.TypeRepr argTys ->
+  ArgShapes ext NoTag argTys ->
+  InitialMem sym ->
+  C.GlobalVar Mem.Mem ->
+  [RefineHeuristic sym bak ext argTys] ->
+  [C.ExecutionFeature p sym ext (C.RegEntry sym ret)] ->
+  ( ( MSM.MacawProcessAssertion sym
+    , Mem.HasLLVMAnn sym
+    ) =>
+    C.GlobalVar ToConc.ToConcretizeType ->
+    Setup.SetupMem sym ->
+    GSIO.InitializedFs sym wptr ->
+    Setup.Args sym ext argTys ->
+    IO (C.ExecState p sym ext (C.RegEntry sym ret))
+  ) ->
+  IO (ProveRefineResult sym ext argTys)
+refineOnce la simOpts halloc bak fm dl valueNames argNames argTys argShapes initMem memVar heuristics execFeats mkInitState = do
+  let sym = C.backendGetSym bak
+  ErrorCallbacks
+    { errorMap = bbMapRef
+    , llvmErrCallback = recordLLVMAnnotation
+    , macawAssertionCallback = processMacawAssert
+    } <-
+    buildErrMaps
+  let ?recordLLVMAnnotation = recordLLVMAnnotation
+  let ?processMacawAssert = processMacawAssert
+  (args, setupMem, setupAnns) <-
+    Setup.setup la bak dl valueNames argTys argShapes initMem
+  initFs_ <- GSIO.initialLlvmFileSystem halloc sym (Opts.simFsOpts simOpts)
+  (toConc, globals1) <- liftIO $ ToConc.newToConcretize halloc (GSIO.initFsGlobals initFs_)
+  let initFs = initFs_{GSIO.initFsGlobals = globals1}
+  st <- mkInitState toConc setupMem initFs args
+  let concInitState =
+        Conc.InitialState
+          { Conc.initStateArgs = args
+          , Conc.initStateFs = GSIO.initFsContents initFs
+          , Conc.initStateMem = initMem
+          }
+  let boundsOpts = Opts.simBoundsOpts simOpts
```

Comment:
it's maybe a bit odd the one can setup refineonce with features that dont implement the bounds declared here, maybe either just pass the timeout or just setup the bounds features here?

Reply:
Good call, done!